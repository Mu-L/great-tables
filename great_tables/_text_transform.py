from __future__ import annotations
from typing import TYPE_CHECKING, Union, List, Callable

from ._gt_data import TextTransformFn, TextTransformFns, TextTransformInfo
from ._tbl_data import SelectExpr
from ._locations import resolve_rows_i, resolve_cols_c

if TYPE_CHECKING:
    from ._types import GTSelf


def text_transform(
    self: GTSelf,
    fns: Union[TextTransformFn, TextTransformFns],
    columns: SelectExpr = None,
    rows: Union[int, List[int], None] = None,
) -> GTSelf:

    # If a single function is supplied to `fns` then
    # repackage that into a list as the `default` function
    if isinstance(fns, Callable):
        fns = TextTransformFns(default=fns)

    row_res = resolve_rows_i(self, rows)
    row_pos = [name_pos[1] for name_pos in row_res]

    col_res = resolve_cols_c(self, columns)

    text_transformer = TextTransformInfo(fns, col_res, row_pos)

    return self._replace(_text_transforms=[*self._text_transforms, text_transformer])
