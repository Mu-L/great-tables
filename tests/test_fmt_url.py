from great_tables import GT
from great_tables.data import towny
from great_tables._utils_render_html import create_body_component_h

towny_mini = towny[["name", "website"]].head(3)


def assert_rendered_body(snapshot, gt):
    built = gt._build_data("html")
    body = create_body_component_h(built)

    assert snapshot == body


def test_fmt_url_01(snapshot):
    new_gt = GT(towny_mini).fmt_url(columns="website")

    assert_rendered_body(snapshot, new_gt)


def test_fmt_url_02(snapshot):
    new_gt = GT(towny_mini).fmt_url(columns="website", as_button=True)

    assert_rendered_body(snapshot, new_gt)


def test_fmt_url_03(snapshot):
    new_gt = GT(towny_mini).fmt_url(columns="website", color="red")

    assert_rendered_body(snapshot, new_gt)


def test_fmt_url_04(snapshot):
    new_gt = GT(towny_mini).fmt_url(columns="website", as_button=True, color="green")

    assert_rendered_body(snapshot, new_gt)


def test_fmt_url_05(snapshot):
    new_gt = GT(towny_mini).fmt_url(columns="website", show_underline=False, color="green")

    assert_rendered_body(snapshot, new_gt)
