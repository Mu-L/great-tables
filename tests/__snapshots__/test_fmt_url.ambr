# serializer version: 1
# name: test_fmt_url_01
  '''
  <tbody class="gt_table_body">
  <tr>
    <td class="gt_row gt_left">Addington Highlands</td>
    <td class="gt_row gt_left"><a style="color:#008B8B;text-decoration:underline;text-underline-position:under;display:inline-block;" href="https://addingtonhighlands.ca">https://addingtonhighlands.ca</a></td>
  </tr>
  <tr>
    <td class="gt_row gt_left">Adelaide Metcalfe</td>
    <td class="gt_row gt_left"><a style="color:#008B8B;text-decoration:underline;text-underline-position:under;display:inline-block;" href="https://adelaidemetcalfe.on.ca">https://adelaidemetcalfe.on.ca</a></td>
  </tr>
  <tr>
    <td class="gt_row gt_left">Adjala-Tosorontio</td>
    <td class="gt_row gt_left"><a style="color:#008B8B;text-decoration:underline;text-underline-position:under;display:inline-block;" href="https://www.adjtos.ca">https://www.adjtos.ca</a></td>
  </tr>
  </tbody>
  '''
# ---
# name: test_fmt_url_02
  '''
  <tbody class="gt_table_body">
  <tr>
    <td class="gt_row gt_left">Addington Highlands</td>
    <td class="gt_row gt_left"><a style="color:#FFFFFF;text-decoration:none;display:inline-block;background-color:#4682B4;padding:8px 12px;;outline-style:solid;outline-color:auto;outline-width:2px;" href="https://addingtonhighlands.ca">https://addingtonhighlands.ca</a></td>
  </tr>
  <tr>
    <td class="gt_row gt_left">Adelaide Metcalfe</td>
    <td class="gt_row gt_left"><a style="color:#FFFFFF;text-decoration:none;display:inline-block;background-color:#4682B4;padding:8px 12px;;outline-style:solid;outline-color:auto;outline-width:2px;" href="https://adelaidemetcalfe.on.ca">https://adelaidemetcalfe.on.ca</a></td>
  </tr>
  <tr>
    <td class="gt_row gt_left">Adjala-Tosorontio</td>
    <td class="gt_row gt_left"><a style="color:#FFFFFF;text-decoration:none;display:inline-block;background-color:#4682B4;padding:8px 12px;;outline-style:solid;outline-color:auto;outline-width:2px;" href="https://www.adjtos.ca">https://www.adjtos.ca</a></td>
  </tr>
  </tbody>
  '''
# ---
# name: test_fmt_url_03
  '''
  <tbody class="gt_table_body">
  <tr>
    <td class="gt_row gt_left">Addington Highlands</td>
    <td class="gt_row gt_left"><a style="color:red;text-decoration:underline;text-underline-position:under;display:inline-block;" href="https://addingtonhighlands.ca">https://addingtonhighlands.ca</a></td>
  </tr>
  <tr>
    <td class="gt_row gt_left">Adelaide Metcalfe</td>
    <td class="gt_row gt_left"><a style="color:red;text-decoration:underline;text-underline-position:under;display:inline-block;" href="https://adelaidemetcalfe.on.ca">https://adelaidemetcalfe.on.ca</a></td>
  </tr>
  <tr>
    <td class="gt_row gt_left">Adjala-Tosorontio</td>
    <td class="gt_row gt_left"><a style="color:red;text-decoration:underline;text-underline-position:under;display:inline-block;" href="https://www.adjtos.ca">https://www.adjtos.ca</a></td>
  </tr>
  </tbody>
  '''
# ---
# name: test_fmt_url_04
  '''
  <tbody class="gt_table_body">
  <tr>
    <td class="gt_row gt_left">Addington Highlands</td>
    <td class="gt_row gt_left"><a style="color:green;text-decoration:none;display:inline-block;background-color:#ADD8E6;padding:8px 12px;;outline-style:none;outline-color:#BEBEBE;outline-width:2px;" href="https://addingtonhighlands.ca">https://addingtonhighlands.ca</a></td>
  </tr>
  <tr>
    <td class="gt_row gt_left">Adelaide Metcalfe</td>
    <td class="gt_row gt_left"><a style="color:green;text-decoration:none;display:inline-block;background-color:#ADD8E6;padding:8px 12px;;outline-style:none;outline-color:#BEBEBE;outline-width:2px;" href="https://adelaidemetcalfe.on.ca">https://adelaidemetcalfe.on.ca</a></td>
  </tr>
  <tr>
    <td class="gt_row gt_left">Adjala-Tosorontio</td>
    <td class="gt_row gt_left"><a style="color:green;text-decoration:none;display:inline-block;background-color:#ADD8E6;padding:8px 12px;;outline-style:none;outline-color:#BEBEBE;outline-width:2px;" href="https://www.adjtos.ca">https://www.adjtos.ca</a></td>
  </tr>
  </tbody>
  '''
# ---
# name: test_fmt_url_05
  '''
  <tbody class="gt_table_body">
  <tr>
    <td class="gt_row gt_left">Addington Highlands</td>
    <td class="gt_row gt_left"><a style="color:green;text-decoration:none;display:inline-block;" href="https://addingtonhighlands.ca">https://addingtonhighlands.ca</a></td>
  </tr>
  <tr>
    <td class="gt_row gt_left">Adelaide Metcalfe</td>
    <td class="gt_row gt_left"><a style="color:green;text-decoration:none;display:inline-block;" href="https://adelaidemetcalfe.on.ca">https://adelaidemetcalfe.on.ca</a></td>
  </tr>
  <tr>
    <td class="gt_row gt_left">Adjala-Tosorontio</td>
    <td class="gt_row gt_left"><a style="color:green;text-decoration:none;display:inline-block;" href="https://www.adjtos.ca">https://www.adjtos.ca</a></td>
  </tr>
  </tbody>
  '''
# ---
